### PR TITLE
experimental 4player Counter / Co-Operative

### PR DESCRIPTION
**Pull request by cylonicboom**
# 4 player pvp missions

Screenshot from 2024-07-05 20-42-26
Screenshot from 2024-07-05 20-46-53

# known issues

- coop + anti is still janky
- dark rooms + NVG or x-ray vision may be a flashing light hazard

# goals

Want to play 4player coop / anti with a mix of coop / anti team characters

No recompiling of ailists. Should "just work" with vanilla levels and well-formed mods.

Load mp profile for coop / anti characters for controller / hud settings / "leveling"

# non-goals

- Custom characters as coop characters (for now)
- Fixing non-breaking n64 bugs beyond what is required
- adding coop / anti support for levels not built for it
- adding anti support to mp ailists commands that don't use it in retail until modders who actually make levels also weigh in on how they should work 

# background

In Vanilla PD,  'coopness' or 'antiness' of a PD session is controlled by setting `g_Vars.coopplayernum` and `g_Vars.antiplayernum` This is used to determine if coop or anti is running, and what player index to use. Also, coop / anti logic is mutually exclusive.

# approach

Porting PD to 4 player local coop/anti requires is set via `g_MissionConfig.isteam` that allows players to join a solo mission as a team of coop or anti characters.

`g_MissionConfig.isteam` is set to `true` when the mission is a team mission.
`g_Vars.playerroles` is used to track player role constants against player slots. This is used by `playermgrAllocatePlayersFromRoles` to initialize the player slots and various control variables. 

`g_Vars.currentplayer` is the current player struct.
`g_Vars.players` : player structs for all players.
`g_Vars.currentplayernum` is the current player number

`g_Vars.coopplayers` : player structs for coop players. The bond slot is set to 0 as well as the anti players.
`g_Vars.currentcoopnum` is the current coop player number. On each frame the coop player is set to the next coop player struct that exists.
`g_Vars.coop`: current coop player struct

`g_Vars.currentantinum` is the current anti player number. On each frame the anti player is set to the next anti player struct that exists
`g_Vars.antiplayers` : player structs for anti players. The anti index is set to 0 as well as the coop players.
`g_Vars.anti`: current anti player struct

In order to gaurentee that ailist commands that are intending to reference `CHR_ANTI`, `CHR_COOP`, and `CHR_P1P2`/`CHR_P1P2O`AIlist commands referencing `CHR_COOP` or `CHR_P1P2` will be treated as any or all coop / ally chaaracters.

# limitations

- scripts targeting `CHR_COOP` or `CHR_P1P2` will need to be treated as any or all ally characters. 
This means that mission-related hud messages intended for one coop character will be sent to them all.

# acknowledgements

- borrowed fixes from @iamgreaser's 4player anti effort: merging the tiktok mode hud settings into 4p, correctly spawning 4 players into a mission
-  @fgsfdsfgs  / Ryan Dwyer for the pc port / decomp but specifically reviewing the scheduling logic and talking me down from attempting any kind of live ailist reinterpretation.

### CODE REVIEW

**Comment by iamgreaser** [comment]
I'm having a look at the ailists for a hopefully cleaner solution and it's starting to look a lot more feasible than I thought it would.

Having said that, having a list of all `anti` and `coop` players is a nice idea. Not sure how tempted I am to borrow that idea, but I do intend to have something working to the point where you'd have co-op and counter-op stuff happening in the same game mode.

Something you'll want to consider is the `CHR_P1P2` reference and the `chr_toggle_p1p2(CHR_SELF)` command. If you cycle through all protagonist players then I suspect that would tackle the most general cases. But there will definitely be plenty of pain especially pertaining to objectives being failed in situations where they should not be, and also stuff appearing and disappearing while visible to some of the extra co-op players.

**Comment by cylonicboom** [comment]
Cycling  the "current coop" player and basing the p1p2 toggle calculation based on MAX_PLAYERS and the current coop number number does make most things work. Like @fgsfdsfgs suggested CHR_COOP / CHR_P1P2_OPPOSITE references in the `if_chr_.*(chr,[... ,] label)` should check for all the other coop characters (when p1p2_opposite isn't bond) so the ailist can make the appropriate jump. Right now non-bond characters need to tap the use button repeatedly for some ailist scripted objects like lifts and objective items because of the coop cycling. Also the surveillance cameras take a long time to notice non-bond players.

**Comment by JacksBestGaming** [comment]
I have no clue who came up with 'Build Your Perfect Team", but I love it so much. This is not only extremely cool on a technical level, but on a creative one as well.

**Comment by Unradelic** [comment]
image

Happened while fiddling with the menu options, going back and forth, then crashed :P

**Comment by cylonicboom** [comment]
> image
> 
> Happened while fiddling with the menu options, going back and forth, then crashed :P

Pretty sure I've seen this while trying to get controllers setup.

**Comment by cylonicboom** [comment]
> image
> 
> Happened while fiddling with the menu options, going back and forth, then crashed :P

https://github.com/fgsfdsfgs/perfect_dark/pull/389/commits/cf75ba31b05d9dbfc45b082f80f0af448b4d585b should fix it if it's the same issue I encountered.

**Comment by cylonicboom** [comment]
- [ ] BUG: vanilla allocation flow disabled and this breaks vanilla campaign flows

**Comment by MonsMarineris** [comment]
Screenshot 2024-09-30 154934
I am unable to build this for some reason :(
I do admit I have no fuggin clue what I'm doing. But still, I was so excited to try this out.

File: src/game/chraicommands.c [hunk 1]
```diff
@@ -5125,22 +5355,45 @@ bool aiRevokeControl(void)
 	struct chrdata *chr = chrFindById(g_Vars.chrdata, cmd[2]);
 
 	if (chr && chr->prop && chr->prop->type == PROPTYPE_PLAYER) {
-		u32 prevplayernum = g_Vars.currentplayernum;
-		u32 playernum = playermgrGetPlayerNumByProp(chr->prop);
-		setCurrentPlayerNum(playernum);
-		bgunSetSightVisible(GUNSIGHTREASON_NOCONTROL, false);
-		bgunSetGunAmmoVisible(GUNAMMOREASON_NOCONTROL, false);
+		if (isChrPropCoop(chr->prop)) {
+			struct playerpool** playerpool = getPlayerPool(CHR_COOP);
```
Comment (cylonicboom) [inline src/game/chraicommands.c:5359]: wtf I have no idea how this built and worked on Linux. I'll take a look when I'm on my computer next. Unblocks https://github.com/fgsfdsfgs/perfect_dark/pull/389#issuecomment-2384121638

**Comment by cylonicboom** [comment]
> Screenshot 2024-09-30 154934
> 
> I am unable to build this for some reason :(
> 
> I do admit I have no fuggin clue what I'm doing. But still, I was so excited to try this out.

I broke windows builds with some stupid code that I'm not sure how it built on Linux. I'll take a look when I'm on my computer next.

**Comment by cylonicboom** [comment]
I'm closing this PR and taking it private until it's in a more stable state.

**Comment by MonsMarineris** [comment]
Thanks for having a look at it. I wish I had any sort of knowledge to help. I'm trying to learn.

**Pull request closed without merging**